service/history/hsm/tree.go

```diff
@@ -380,6 +391,54 @@ func (n *Node) InternalRepr() *persistencespb.StateMachineNode {
 	return n.persistence
 }
 
+// Sync updates the current node with the incoming node if state in the incoming node is newer.
+// Meant to be used by the framework, **not** by components.
+func (n *Node) Sync(incomingNode *Node) (bool, error) {
+	currentState, err := MachineData[any](n)
+	if err != nil {
+		return false, err
+	}
+	incomingState, err := MachineData[any](incomingNode)
+	if err != nil {
+		return false, err
+	}
+
+	if result, err := n.definition.CompareState(currentState, incomingState); err != nil || result >= 0 {
```

Comment:
I'm thinking we may not need this if we also store the last update namespace failover version and never increment transition count on the passive side, instead we could sync the node (machine) transition count from the active cluster.

With the exception of history event apply lag (which should be accounted for by comparing the version histories), the only time when version + machine transition count would diverge is on split brain. And in that case we need to overwrite our state with the one from the winning cluster - which would rely on version history + the last updated node's namespace failover version to break the tie.

You'd also want to avoid using `MachineTransition` below and instead sync over the node state.

Reply:
Syned offline. Trying to make machine transition count in sync across cluster is very tricky with existing replication stack. To make that work, event resend logic needs to account for state only changes as well and that would be too much for the short term nexus replication solution.